Use native Ginkgo test runner instead of cmd/e2e

ixdy opened the pull request:
This PR deletes cmd/e2e and updates hack/ginkgo-e2e.sh to use the `ginkgo` command instead. All logic from cmd/e2e/e2e.go and test/e2e/driver.go have been combined into the new file test/e2e/e2e_test.go.

Additionally, several tests which made poor assumptions about cwd or used testContext before it was set have been fixed.

This change is generally intended to have no externally visible changes,
aside from the following caveats:
- Developers must run `go get github.com/onsi/ginkgo/ginkgo` once
- The `-t`/`--tests`, `--times`, and `--orderseed` flags have been removed (such functionality is already provided by Ginkgo itself)
- Calling cmd/e2e/e2e directly obviously won't work, but that was never
  supported anyway
- If the `GINKGO_PARALLEL` environment variable is set to `y`, then ginkgo will run test specs in pa rallel. (Currently defaults to `n`, since some tests are broken in this mode.)

Fixes #5670.

ixdy committed:
Use native Ginkgo test runner instead of cmd/e2e.

This commit deletes cmd/e2e and updates hack/ginkgo-e2e.sh to use the
'ginkgo' command instead. All logic from cmd/e2e/e2e.go and
test/e2e/driver.go have been combined into the new file
test/e2e/e2e_test.go.

Additionally, several tests which made poor assumptions about cwd or
used testContext before it was set have been fixed.

This change is generally intended to have no externally visible changes,
aside from the following caveats:
- The -t/--tests flag has been removed
- Calling cmd/e2e/e2e directly obviously won't work, but that was never
  supported anyway
- If the GINKGO_PARALLEL environment variable is set to y, then ginkgo
  will run test specs in parallel. (Currently defaults to n, since some
  tests are broken in this mode.)

ixdy committed:
Call the cluster/kubectl.sh wrapper in e2e tests.

The ginkgo wrapper script used to set up the path so that tests would
find the appropriate kubectl binary, but this seems like a cleaner way
to achieve the same result.

ixdy commented:
@quinton-hoole @zmerlynn 

I'll note that I'm not particularly happy about requiring a `go get github.com/onsi/ginkgo/ginkgo` to be run manually. There's probably some way to get the `ginkgo` binary built as part of the rest of the build, but I'm not really sure how to get our "build system" to build stuff in Godeps etc and then reference that from our test scripts. I'm welcome to suggestions/feedback, though.

ixdy commented:
I figured out how to use godep to build the checked-in version of Ginkgo and use it each time.

quinton-hoole commented:
Excellent Jeff.  I'll start reviewing shortly.

ixdy commented:
I've run e2e tests locally (via `go run hack/e2e.go -v --test` or `hack/ginkgo-e2e.sh`). They are still passing.

ixdy commented:
Updated with a few fixes to ensure that Godeps are used when building the e2e tests.

Additionally, I added support to run Ginkgo in parallel mode (with `GINKGO_PARALLEL=y`), though this is turned off by default.

zmerlynn commented on test/e2e/e2e_test.go:
```diff
@@ -0,0 +1,127 @@
+var (
+	cloudConfig = &testContext.CloudConfig
+
+	orderseed = flag.Int64("orderseed", 0, "If non-zero, seed of random test shuffle order. (Otherwise random.)")
```
I don't think this will ever be needed again. The gomega command already has support for a fixed seed.

ixdy replied:
good point. I'll remove it, along with its corresponding TODO.

zmerlynn commented on test/e2e/e2e_test.go:
```diff
@@ -0,0 +1,127 @@
+
+	orderseed = flag.Int64("orderseed", 0, "If non-zero, seed of random test shuffle order. (Otherwise random.)")
+	reportDir = flag.String("report-dir", "", "Path to the directory where the JUnit XML reports should be saved. Default is empty, which doesn't generate these reports.")
+	times     = flag.Int("times", 1, "Number of times each test is eligible to be run. Individual order is determined by shuffling --times instances of each test using --orderseed (like a multi-deck shoe of cards).")
```
Do we still need this?

ixdy replied:
I didn't remove it because I didn't want to remove any functionality in this change, but I'm fine to remove it if we think it isn't worthwhile to keep. (Ginkgo has a builtin --untilItFails option which will keep running until tests fail, which may be a better option anyway.)

zmerlynn replied:
I actually don't mind removing it because of `--untilItFails`, because that's the thing I'm usually trying to do if I want to run a set of tests multiple times. Also because it decreases the size of the entry code here, which looks like something we'd have to share with each "suite" - unless there's a different way to organize that you see? (After moving to the Ginkgo runner, I'd really like to pick off the density/long tests and get them out of e2e/hackiness, and I think it can be done with a minor refactor here.)

zmerlynn commented on test/e2e/e2e_test.go:
```diff
@@ -0,0 +1,127 @@
+func TestE2E(t *testing.T) {
+	defer util.FlushLogs()
+
+	// Disable density test unless it's explicitly requested.
```
I assume after this we can actually just split density off into its own directory and stop skipping it this BS way?

ixdy replied:
Yes, we should be easily able to split off tests into their own packages and then choose which tests to run via the cmdline - we could have test/e2e/density or whatever which will include only those tests. (Would need a separate density_test.go file, in that case, so we might need to refactor some of the code here in e2e_test.go, but very feasible.)

zmerlynn commented:
This is awesome. Now that the release is cut and we're a little more stable, let's maybe get this in tomorrow morning?

ixdy committed:
Remove --orderseed and --times flags from the e2e test.

Similar functionality for both is already provided by the Ginkgo runner
itself.

ixdy commented:
OK, I added a new commit which removes `--times` and `--orderseed`. (Separate commit so you can see what changed, but I'm happy to squash, too.)

I left in the density hackiness and AWS stuff for now, though these should probably also get cleaned up at some point (maybe not in this change though?).

zmerlynn commented:
Yeah, no problem with incremental changes. This change alone is pretty breaking and I just want to see it soak. It's also pretty enabling given the parallel e2e.

Have you figured out how to block parallelism at all?

zmerlynn commented:
(Those buttons are too damn close.)

ixdy commented:
Note that as currently written, it's not particularly easy to pass `--untilItFails` since this needs to be passed to the ginkgo runner binary, not to the e2e tests - similar to what I needed to do for parallel tests. I can add another argument to hack/ginkgo-e2e.sh, or I can leave it as is for now...

What do you mean by "block parallelism"? By default this still won't run tests in parallel.

zmerlynn commented:
Just thinking for later. We're going to need a way to indicate that some tests can't run in parallel, and I didn't know if Ginkgo already had that facility. The way may be a sledgehammer, though: We could just organize it so that `services.sh`, our last real holdout that I *know* won't work, goes in a "long" suite, along with the reboot test that @mbforbes is working on (for example), etc., and that stuff just never gets executed in parallel.

zmerlynn commented:
But this all LGTM. I'd prefer to wait 'til morning.

ixdy commented:
Oddly enough, in my limited testing of parallel mode, it seemed like `services.sh` was doing just fine, though I don't have 100% confidence in that. (There were a few other tests which seemed less parallel-safe, likely due to not using namespaces.)

Overall I think we need a better strategy for selecting which tests to run. For example, all of the manual checking of cloud provider is an antipattern that causes bugs like #7008, and none of Ginkgo's metadata even indicates that such tests are skipped.

ixdy commented:
@justinsb FYI, this shouldn't break your Jenkins instance, but it might be worth keeping an eye on as soon as it gets merged.

a-robinson commented:
> The -t/--tests, --times, and --orderseed flags have been removed (such functionality is already provided by Ginkgo itself)

Which ginkgo flag replaces --times? For some reason the --times flag kept working for me until today, and I don't see anything related in the list of flags spit out when it complains about the unknown flag.

ixdy commented:
Ginkgo has `--untilItFails`, which is similar to the functionality (usually) desired by `--times`.

a-robinson commented:
I apparently have no idea how to run e2e tests anymore, since that flag also isn't recognized. This is what I used to run (as a modified hack/e2e-test.sh):
```
go run "$(dirname $0)/e2e.go" -v -test --check_version_skew=false --test_args="--ginkgo.focus=Services.*external* --times=10"
```
This is one of the variations I've tried just now:
```
go run "$(dirname $0)/e2e.go" -v -test --check_version_skew=false --test_args="--ginkgo.focus=Services.*external* --untilItFails=true"
```
It gives the error `flag provided but not defined: -untilItFails`. What should I be running to make this work?

ixdy commented:
Oh, it seems like it needs to be specified as one of the args in $ginkgo_args in hack/ginkgo-e2e.sh. We probably need to allow adding these with an environment variable or something.
